File: app/src/main/java/com/gonodono/appiconchangedemo/IconChangeManager.java
package com.gonodono.appiconchangedemo;

import static android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DEFAULT;
import static android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
import static android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_ENABLED;
import static android.content.pm.PackageManager.DONT_KILL_APP;
import static android.content.pm.PackageManager.GET_ACTIVITIES;
import static android.content.pm.PackageManager.GET_DISABLED_COMPONENTS;
import static android.content.pm.PackageManager.GET_META_DATA;

import android.app.Activity;
import android.content.ComponentName;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.List;

class IconChangeManager {

    private static final String
            META_DATA_KEY_ALIAS_TITLE = BuildConfig.APPLICATION_ID + ".ALIAS_TITLE";
    private static final String
            EXTRA_REFRESHING = BuildConfig.APPLICATION_ID + ".EXTRA_REFRESHING";

    private final Activity activity;
    private final PackageManager packageManager;
    private final List<Alias> iconAliases;
    private final Alias initialAlias;
    private final Alias cloneInitialAlias;
    private final Alias currentAlias;
    private final boolean isIconChangeActivated;

    private boolean isRefreshing;

    IconChangeManager(Activity activity) {
        this.activity = activity;
        this.packageManager = activity.getPackageManager();
        this.iconAliases = new ArrayList<>();

        Alias initial = null;
        Alias cloneInitial = null;
        Alias current = null;
        boolean isActivated;

        try {
            final PackageInfo packageInfo = getPackageInfo();
            for (ActivityInfo activityInfo : packageInfo.activities) {
                if (isIconChangeActivityAlias(activityInfo)) {
                    final Alias alias =
                            new Alias(activityInfo.name,
                                    activityInfo.metaData.getString(META_DATA_KEY_ALIAS_TITLE),
                                    activityInfo.icon);

                    if (isAliasComponentEnabled(activityInfo)) {
                        if (current == null) {
                            current = alias;
                        } else {
                            throw new IllegalStateException("Multiple aliases currently enabled");
                        }
                    }

                    if (activityInfo.enabled) {
                        if (initial == null) {
                            initial = alias;
                        } else {
                            throw new IllegalStateException(
                                    "Multiple initial aliases enabled in manifest");
                        }
                    } else {
                        iconAliases.add(alias);
                    }
                }
            }

            if (current == null) {
                throw new IllegalStateException("No alias currently enabled");
            }

            if (initial == null) {
                throw new IllegalStateException("No initial alias enabled in manifest");
            } else {
                isActivated = packageManager
                        .getComponentEnabledSetting(new ComponentName(activity, initial.className))
                        == COMPONENT_ENABLED_STATE_DISABLED;
            }

            for (Alias alias : iconAliases) {
                if (!alias.equals(initial) && alias.iconResId == initial.iconResId) {
                    if (cloneInitial == null) {
                        cloneInitial = alias;
                    } else {
                        throw new IllegalStateException("Multiple clone initial aliases found");
                    }
                }
            }

            if (cloneInitial == null) {
                throw new IllegalStateException("No clone initial alias found");
            }
        } catch (PackageManager.NameNotFoundException e) {
            throw new RuntimeException(e);
        }

        initialAlias = initial;
        cloneInitialAlias = cloneInitial;
        currentAlias = current;
        isIconChangeActivated = isActivated;
    }

    List<Alias> getAliases() {
        return iconAliases;
    }

    Alias getCurrentAlias() {
        return currentAlias;
    }

    boolean isIconChangeActivated() {
        return isIconChangeActivated;
    }

    void activateIconChange() {
        if (!isIconChangeActivated) {
            setAliasComponentEnabled(cloneInitialAlias, true);
            setAliasComponentEnabled(initialAlias, false);
            activity.finish();
        }
    }

    boolean deactivateIconChange() {
        if (isIconChangeActivated) {
            try {
                final PackageInfo packageInfo = getPackageInfo();
                for (ActivityInfo activityInfo : packageInfo.activities) {
                    if (isIconChangeActivityAlias(activityInfo)) {
                        packageManager.setComponentEnabledSetting(
                                createComponentName(activityInfo.name),
                                COMPONENT_ENABLED_STATE_DEFAULT,
                                DONT_KILL_APP);
                    }
                }
                refresh();
                return true;
            } catch (Exception e) {
                e.printStackTrace();
                return false;
            }
        } else {
            return false;
        }
    }

    void setCurrentAlias(Alias alias) {
        if (!currentAlias.equals(alias)) {
            setAliasComponentEnabled(alias, true);
            setAliasComponentEnabled(currentAlias, false);
            refresh();
        }
    }

    void onSaveInstanceState(Bundle outState) {
        outState.putBoolean(EXTRA_REFRESHING, isRefreshing);
    }

    boolean isIconChangeRefresh(Bundle savedInstanceState) {
        return savedInstanceState != null && savedInstanceState.getBoolean(EXTRA_REFRESHING);
    }

    private PackageInfo getPackageInfo() throws PackageManager.NameNotFoundException {
        return packageManager.getPackageInfo(activity.getPackageName(),
                GET_ACTIVITIES | GET_DISABLED_COMPONENTS | GET_META_DATA);
    }

    private boolean isIconChangeActivityAlias(ActivityInfo activityInfo) {
        return activityInfo.targetActivity != null && activityInfo.metaData != null &&
                activityInfo.metaData.containsKey(META_DATA_KEY_ALIAS_TITLE);
    }

    private boolean isAliasComponentEnabled(ActivityInfo aliasInfo) {
        final int state =
                packageManager.getComponentEnabledSetting(createComponentName(aliasInfo.name));
        if (state == COMPONENT_ENABLED_STATE_DEFAULT) {
            return aliasInfo.enabled;
        }
        return state == COMPONENT_ENABLED_STATE_ENABLED;
    }

    private void setAliasComponentEnabled(Alias alias, boolean enabled) {
        final int newState =
                enabled ? COMPONENT_ENABLED_STATE_ENABLED :
                        alias.equals(initialAlias) ? COMPONENT_ENABLED_STATE_DISABLED :
                                COMPONENT_ENABLED_STATE_DEFAULT;

        packageManager.setComponentEnabledSetting(createComponentName(alias.className),
                newState, DONT_KILL_APP);
    }

    private ComponentName createComponentName(String className) {
        return new ComponentName(activity.getPackageName(), className);
    }

    private void refresh() {
        isRefreshing = true;
        activity.recreate();
    }
}

File: app/src/main/kotlin/com/gonodono/appiconchangedemo/IconChangeManager.kt
package com.gonodono.appiconchangedemo

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.content.pm.ActivityInfo
import android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DEFAULT
import android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DISABLED
import android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_ENABLED
import android.content.pm.PackageManager.DONT_KILL_APP
import android.content.pm.PackageManager.GET_ACTIVITIES
import android.content.pm.PackageManager.GET_META_DATA
import android.content.pm.PackageManager.MATCH_DISABLED_COMPONENTS
import android.os.Bundle

internal class IconChangeManager(private val activity: Activity) {

    private val currentInit: ActivityAlias
    private val initialAlias: ActivityAlias
    private val cloneInitialAlias: ActivityAlias

    var selectableAliases: List<ActivityAlias>
        private set

    init {
        val aliases = activity.getActivitiesFromManifest()
            .filter { activityInfo -> activityInfo.isIconChangeAlias }
            .map { activityInfo -> ActivityAlias(activityInfo) }

        val componentEnabled = aliases.filter { it.isComponentEnabled }
        currentInit = when (componentEnabled.size) {
            1 -> componentEnabled[0]
            0 -> error("No alias component currently enabled")
            else -> error("Multiple alias components currently enabled")
        }

        val (enabled, disabled) = aliases.partition { it.isEnabledInManifest }
        initialAlias = when (enabled.size) {
            1 -> enabled[0]
            0 -> error("No initial alias enabled in manifest")
            else -> error("Multiple aliases enabled in manifest")
        }

        val sameIcon = disabled.filter { it.icon == initialAlias.icon }
        cloneInitialAlias = when (sameIcon.size) {
            1 -> sameIcon[0]
            0 -> error("No clone initial alias found")
            else -> error("Multiple clone initial aliases found")
        }

        selectableAliases = disabled
    }

    var currentAlias: ActivityAlias = currentInit
        set(alias) {
            if (field == alias) return
            field.isComponentEnabled = false
            alias.isComponentEnabled = true
            field = alias
            isChangingIcon = true
            activity.recreate()
        }

    private var isChangingIcon = false

    fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean(EXTRA_IS_CHANGING_ICON, isChangingIcon)
    }

    var isIconChangeActivated: Boolean = !initialAlias.isComponentEnabled
        set(activated) {
            if (field == activated) return
            field = activated
            if (activated) {
                initialAlias.isComponentEnabled = false
                cloneInitialAlias.isComponentEnabled = true

                // Originally, the demo just finished the Activity and required
                // the user to relaunch. While updating, I found that a restart
                // works too, at least with this simple setup. If you have any
                // issues, you might need to change back to a manual relaunch.
                activity.restart()
            } else {
                initialAlias.isComponentEnabled = true
                currentAlias.isComponentEnabled = false

                // Activity#recreate() seems to suffice here as well, since the
                // aliases are being reset to their default values, but the demo
                // does a restart to keep things symmetrical. You can pull this
                // out of the if-else, obviously, if you keep the restarts.
                activity.restart()
            }
        }

    enum class StartMode { Normal, ActivationChange, IconChange }

    fun determineStartMode(
        intent: Intent,
        savedInstanceState: Bundle?
    ): StartMode = when {
        // Order matters, unless you'd prefer to reset the Intent extra below.
        savedInstanceState?.getBoolean(EXTRA_IS_CHANGING_ICON) ?: false -> {
            StartMode.IconChange
        }
        intent.getBooleanExtra(EXTRA_IS_CHANGING_ACTIVATION, false) -> {
            StartMode.ActivationChange
        }
        else -> StartMode.Normal
    }

    private var ActivityAlias.isComponentEnabled: Boolean
        get() {
            val manager = activity.packageManager
            val component = ComponentName(activity, name)
            val state = manager.getComponentEnabledSetting(component)
            return if (state == COMPONENT_ENABLED_STATE_DEFAULT) {
                isEnabledInManifest
            } else {
                state == COMPONENT_ENABLED_STATE_ENABLED
            }
        }
        set(enabled) {
            val manager = activity.packageManager
            val component = ComponentName(activity, name)
            val state = if (this == initialAlias) {
                if (enabled) COMPONENT_ENABLED_STATE_DEFAULT
                else COMPONENT_ENABLED_STATE_DISABLED
            } else {
                if (enabled) COMPONENT_ENABLED_STATE_ENABLED
                else COMPONENT_ENABLED_STATE_DEFAULT
            }
            manager.setComponentEnabledSetting(component, state, DONT_KILL_APP)
        }
}

private fun Activity.getActivitiesFromManifest(): Array<out ActivityInfo> {
    val flags = GET_ACTIVITIES or GET_META_DATA or MATCH_DISABLED_COMPONENTS
    return packageManager.getPackageInfo(packageName, flags).activities!!
}

private val ActivityInfo.isIconChangeAlias: Boolean
    get() = metaData?.containsKey(ActivityAlias.META_DATA_TITLE) == true

private fun Activity.restart() {
    finish() // <- Must come first, else it cancels the startActivity().
    val intent = packageManager.getLaunchIntentForPackage(packageName)!!
    startActivity(intent.putExtra(EXTRA_IS_CHANGING_ACTIVATION, true))
}

private const val EXTRA_IS_CHANGING_ACTIVATION: String =
    BuildConfig.APPLICATION_ID + ".extra.IS_CHANGING_ACTIVATION"

private const val EXTRA_IS_CHANGING_ICON: String =
    BuildConfig.APPLICATION_ID + ".extra.IS_CHANGING_ICON"